trace-dispatcher/src/Cardano/Logging/ConfigurationParser.hs

```diff
@@ -19,53 +20,108 @@ import qualified Data.ByteString.Char8 as BS
 import           Data.List (foldl')
 import qualified Data.Map.Strict as Map
 import           Data.Maybe (isJust)
-import           Data.Text (Text, split)
+import           Data.Text (Text, intercalate, split)
 import           Data.Yaml
 import           GHC.Generics
 
-
 import           Cardano.Logging.Types
 
-defaultConfig :: TraceConfig
-defaultConfig = emptyTraceConfig {
-  tcOptions = Map.fromList
-    [([] :: [Text],
-         [ ConfSeverity (SeverityF (Just Info))
-         , ConfDetail DNormal
-         , ConfBackend [Stdout HumanFormatColoured]
-         ])
-    ]
-  }
-
 -- -----------------------------------------------------------------------------
 -- Configuration file
 
+-- | The external representation of a configuration file
+data ConfigRepresentation = ConfigRepresentation {
+    traceOptions                  :: OptionsRepresentation
+  , traceOptionForwarder          :: Maybe TraceOptionForwarder
+  , traceOptionNodeName           :: Maybe Text
+  , traceOptionPeerFrequency      :: Maybe Int
+  , traceOptionResourceFrequency  :: Maybe Int
+  }
+  deriving (Eq, Ord, Show)
+
+instance AE.FromJSON ConfigRepresentation where
+    parseJSON (Object obj) = ConfigRepresentation
+                           <$> obj .: "TraceOptions"
+                           <*> obj .:? "TraceOptionForwarder"
+                           <*> obj .:? "TraceOptionNodeName"
+                           <*> obj .:? "TraceOptionPeerFrequency"
+                           <*> obj .:? "TraceOptionResourceFrequency"
+    parseJSON _ = mempty
+
+instance AE.ToJSON ConfigRepresentation where
+  toJSON ConfigRepresentation{..} = object
+    [ "TraceOptions"                  .= traceOptions
+    , "TraceOptionForwarder"          .= traceOptionForwarder
+    , "traceOptionNodeName"           .= traceOptionNodeName
+    , "TraceOptionPeerFrequency"      .= traceOptionPeerFrequency
+    , "traceOptionResourceFrequency"  .= traceOptionResourceFrequency
+    ]
+
+-- | In the external configuration representation for configuration files
+-- all options for a namespace are part of a record
+data ConfigOptionRep = ConfigOptionRep
+    { severity :: Maybe SeverityF
+    , detail :: Maybe DetailLevel
+    , backends :: Maybe [BackendConfig]
+    , maxFrequency :: Maybe Double
+    }
+  deriving (Eq, Ord, Show, Generic)
+
+instance AE.FromJSON ConfigOptionRep where
```

Comment:
Is there any reason why `deriving (FromJSON, ToJSON)` isn't used? I can't really spot a difference in the handcrafted instances.

Reply:
The maybe fields are only printed, when the value is not Nothing. This is hand coded, so that I used handcrafted instances here. I think this is only possible with Template Haskell automatically, but in my opinion the handwritten stuff is preferred.